packages/framework/src/Component/Money/Money.php

```diff
@@ -4,9 +4,10 @@
 
 namespace Shopsys\FrameworkBundle\Component\Money;
 
+use JsonSerializable;
 use Litipk\BigNumbers\Decimal;
 
-class Money
+class Money implements JsonSerializable
```

Comment:
It helped us solve the issue with JS validation, it was pretty easy to do and it is [consistent with the Money type](https://github.com/moneyphp/money/blob/master/src/Money.php#L512-L523). What would you suggest to do instead?

Reply:
Doing serialization out of the `Money` eg.:

```php
class MoneyJsonSerializer
{
  public function serialize(Money $money): array {
    return ['amount' => $money->toString()];
  }
}

Reply:
It could be rewritten like this:

```diff
diff --git a/packages/framework/src/Component/Money/Money.php b/packages/framework/src/Component/Money/Money.php
index ef3150a29..e0bc937ef 100644
--- a/packages/framework/src/Component/Money/Money.php
+++ b/packages/framework/src/Component/Money/Money.php
@@ -4,10 +4,9 @@ declare(strict_types=1);
 
 namespace Shopsys\FrameworkBundle\Component\Money;
 
-use JsonSerializable;
 use Litipk\BigNumbers\Decimal;
 
-class Money implements JsonSerializable
+class Money
 {
     /**
      * @var \Litipk\BigNumbers\Decimal
@@ -70,16 +69,6 @@ class Money implements JsonSerializable
         return (string)$this->decimal;
     }
 
-    /**
-     * @return string[]
-     */
-    public function jsonSerialize(): array
-    {
-        return [
-            'amount' => $this->toString(),
-        ];
-    }
-
     /**
      * @param \Shopsys\FrameworkBundle\Component\Money\Money $money
      * @return \Shopsys\FrameworkBundle\Component\Money\Money
diff --git a/packages/framework/src/Form/JsFormValidatorFactory.php b/packages/framework/src/Form/JsFormValidatorFactory.php
index 83bcdddb9..336a55468 100644
--- a/packages/framework/src/Form/JsFormValidatorFactory.php
+++ b/packages/framework/src/Form/JsFormValidatorFactory.php
@@ -5,6 +5,7 @@ namespace Shopsys\FrameworonekBundle\Form;
 use Fp\JsFormValidatorBundle\Factory\JsFormValidatorFactory as BaseJsFormValidatorFactory;
 use Fp\JsFormValidatorBundle\Model\JsFormElement;
 use JsonSerializable;
+use Shopsys\FrameworkBundle\Component\Money\Money;
 use Shopsys\FrameworkBundle\Model\Product\Parameter\ParameterValue;
 use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
 use Symfony\Component\Form\Form;
@@ -85,7 +86,7 @@ class JsFormValidatorFactory extends BaseJsFormValidatorFactory
         $model = parent::createJsModel($form);
 
         if ($model !== null) {
-            $this->jsonSerializeValuesInAllConstraints($model);
+            $this->jsonSerializeMoneyValuesInAllConstraints($model);
         }
 
         return $model;
@@ -106,21 +107,20 @@ class JsFormValidatorFactory extends BaseJsFormValidatorFactory
     }
 
     /**
-     * Method searches for all Constraints and serializes all their values implementing JsonSerializable
-     * (eg. $min and $max in {@see \Shopsys\FrameworkBundle\Form\Constraints\MoneyRange}) because
-     * {@see \Fp\JsFormValidatorBundle\Model\JsModelAbstract::phpValueToJs()} does not support JsonSerializable objects
+     * Method searches for all Constraints and serializes all their values of Money type
+     * (eg. $min and $max in {@see \Shopsys\FrameworkBundle\Form\Constraints\MoneyRange})
      *
      * Method does not modify the original Constraint objects (it clones all the Constraints)
      *
      * @param \Fp\JsFormValidatorBundle\Model\JsFormElement $model
      */
-    protected function jsonSerializeValuesInAllConstraints(JsFormElement $model): void
+    protected function jsonSerializeMoneyValuesInAllConstraints(JsFormElement $model): void
     {
         if (isset($model->data['form']['constraints'])) {
             foreach ($model->data['form']['constraints'] as $constraintName => $constraintSet) {
                 foreach ($constraintSet as $key => $constraint) {
                     $constraint = clone $constraint;
-                    $this->jsonSerializeConstraintValues($constraint);
+                    $this->jsonSerializeConstraintMoneyValues($constraint);
 
                     $model->data['form']['constraints'][$constraintName][$key] = $constraint;
                 }
@@ -131,11 +131,13 @@ class JsFormValidatorFactory extends BaseJsFormValidatorFactory
     /**
      * @param \Symfony\Component\Validator\Constraint $constraint
      */
-    protected function jsonSerializeConstraintValues(Constraint $constraint): void
+    protected function jsonSerializeConstraintMoneyValues(Constraint $constraint): void
     {
         foreach ($constraint as $name => $value) {
-            if ($value instanceof JsonSerializable) {
-                $constraint->$name = $value->jsonSerialize();
+            if ($value instanceof Money) {
+                $constraint->$name = [
+                    'amount' => $value->toString(),
+                ];
             }
         }
     }

```

But we've discussed it with @TomasLudvik and there is definitely added value in `JsFormValidatorFactory` understanding the concept of `JsonSerializable` objects in `Constraint`s, if we want to use value objects in our code more. Furthermore there might be added value in `Money` being able to serialize to a normalized format and JSON seems quite suitable.

Money type in `moneyphp/money` uses this approach too.

I'm undecided here, to be honest. What do others think?

Reply:
As I already said couple of times - it is mixing responsibilities for me; and if we are able to separate those responsibilities (and we are, pretty easily), then I'm always for SRP.

Reply:
From my point of view (correct me if I'm wrong), serialization is currently used mainly for constraints. But that doesn't mean it cannot be used anywhere else in the future.

At first, it seemed logical to me for `Money` to implement `JsonSerializable` but the more I think about it the more it looks like a bad idea. Not just because of the violation of SRP (and here I really don't know if it is the violation or not. The line seems a little blurry), but we can set the path for potential god-object. 

If I want to "transform" Money in some different way (pretty print, to XML, ...) I always create a separate class with separate responsibility. And thinking about that, I believe it should be done the same for JSONs.

And I would really appreciate the class responsible for serialization (even if it would be used in just one place) instead of creating the array in place. But that's just my opinion and I'm not forcing anything :)

Reply:
I don't see any problem with Money being `JsonSerializable`. JSON serialization is mainly used when you need to work with forms on frontend (JS validation, sending with AJAX, ...) or when you need to get some data from backend (again with AJAX). I see that the case of Money is almost the same. Some transformer would work but I don't see the added value of it, because for pretty printing and XML's we mostly use twig filters.
But i'm okay with both outcomes.

Reply:
The current `Money` responsibility is not simple - it is to represent money in the shop. It has to be always consistent, it must greatly cover edge cases, it must handle money operations. The `Money` becomes a stable piece of the SSFW, it is something we rely on.

The `JsonSerializable` suggest to add a method for JSON serialization. JSON serialization is not mentioned in the Money responsibility explained earlier. JSON serialization is not part of "working with money" domain.

Also JSON serialization is used only out of the model - in forms; maybe ajax; maybe admin - these use cases are not model use cases, therefore methods for these use cases should not be part of the model.

Reply:
Advantages of the two approaches:

Using `JsonSerializable`:
- easiest usage of the serialization
- changes in `JsFormValidatorFactory` can be immediately reused for any `JsonSerializable` constraint parameter
- same approach as in `moneyphp/money`

Using separate class for serializing Money to JSON:
- avoids mixing responsibilities
- usage can be found more easily in IDE (Money can be passed to any `json_encode` call)
- can implement deserialization as well as serialization (not used in Shopsys Framework atm)

The serialization is used at two places in the application: `JsFormValidatorFactory` (constraint parameters) and `Admin/Content/Order/orderItem.html.twig` (transport and payment prices in array).